spring-cloud-aws-sqs/src/main/java/io/awspring/cloud/sqs/CompletableFutures.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2013-2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.awspring.cloud.sqs;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+/**
+ * Utility methods for using convenient {@link CompletableFuture} methods from later JDK versions in Java 8.
+ *
+ * @author Tomaz Fernandes
+ * @since 3.0
+ */
+public class CompletableFutures {
```

Comment:
I wonder if those methods hide readability on the `CompletableFutures` code somehow

Reply:
Not sure what you mean by that... These methods are mainly methods that have been added to the `CompletableFuture` API in later JDK versions (9+).

IMHO they're really handy as they make some of the complexity of `CompletableFutures` more expressive.

I think when it comes to the `CompletableFuture` API any help is welcome 😄

WDYT?

Reply:
I see your point here. The tradeoff is between optimizing for figuring out `what` the code does, as opposed to `how` the code does.

While the `CompletableFutures` and `ConfigUtil` classes improves the former, it makes the latter more difficult.

IMHO the `ConfigUtil` class' methods are straightforward enough so that we can safely optimize to `what`, and if necessary a quick look into the class should elucidate any doubts as to `how`.

In `CompletableFutures` this is more of a though decision. I think being able to figure out quickly `what` a `CompletableFuture` chain does is a big advantage and this class helps this a lot by removing unnecessary repetitive code from the chains, and making it more expressive.

OTOH, when figuring out `how` the chain is organized is necessary, i.e. for debugging, having the chain upfront makes the task easier.

In general, I tend to think we'll read the code much more often than debug it, so in general we should optimize for the `what` rather than the `how`. In this case, when debugging we can always make the chain explicit if we want to and rollback after we found the issue (just did it now with your point in mind).

I think ideally we'd try both approaches and see how it looks. In my experience with `CompletableFutures` so far, the extra readability has been worth it in keeping overall chain complexity down for readability.

Let me know if you have any further thoughts on this, and thanks for pointing this out.